### PR TITLE
Adding FAQ entry for elastic search

### PR DESCRIPTION
**Pull request by mbanikazemi**
Creating a new category of FAQs for application specific configuration
and adding the first entry for Elasticsearch

Addresses https://github.com/istio/istio/issues/14743

Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by linsun** [comment]
General comment - is this FAQ entry for permissive mode only?  What about strict mTLS?

File: content/en/faq/applications/elasticsearch.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
+---
+title: Elasticsearch Configuration
```
Comment (frankbu) [inline content/en/faq/applications/elasticsearch.md]: FAQ entries should be in the form of a question

File: content/en/faq/applications/elasticsearch.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+description: Elasticsearch configuration required for use with Istio
+weight: 50
+keywords: [elasticsearch]
+---
+
+1. Elasticsearch Configuration
+
+There are two Elasticsearch configuration parameters that need to be
+set appropriately for running Elasticsearch with Istio:
```
Comment (frankbu) [inline content/en/faq/applications/elasticsearch.md]: Why do they need to be set appropriately? What is the problem? I think the FAQ entry should start by saying something like "Depending on how it's configured, Elasticsearch may or may not work properly when running in Istio". Then explain what the problem is and that a common example is when the network.host parameter is configured to the podIP value. If instead, it's left unset (default value is 0.0.0.0), or explicitly set to 0.0.0.0 or localhost, then it should work without any further specialized configuration.

File: content/en/faq/applications/elasticsearch.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html#modules-network)
+for more information.
+
+With some configurations, ES can be deployed and work correctly
+without a service mesh while failing to work in a service mesh.  A
+commonly seen such configuration sets the `network.host` parameter to
+the pod IP. In cases where Elasticsearch dose not pick the correct
+addresses by default, for example when the host has multiple IP
+addresses, the `network.publish_host` and `network.bind_host`
+parameter should be set to the desired IP addresses.
```
[the remark this answers is not in the document]
Comment (frankbu) [reply content/en/faq/applications/elasticsearch.md]: Agree this is confusing. How does the user know when to do this? Is it just in the case that implicit or explicit 0.0.0.0 network.host didn't work? If so, what's the root problem in this case?

File: content/en/faq/applications/elasticsearch.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+{{< text plain >}}
```
Comment (frankbu) [inline content/en/faq/applications/elasticsearch.md]:
```suggestion
{{< text yaml >}}
```

File: content/en/faq/applications/elasticsearch.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+for more information.
+
+With some configurations, ES can be deployed and work correctly
+without a service mesh while failing to work in a service mesh.  A
+commonly seen such configuration sets the `network.host` parameter to
+the pod IP. In cases where Elasticsearch dose not pick the correct
+addresses by default, for example when the host has multiple IP
+addresses, the `network.publish_host` and `network.bind_host`
+parameter should be set to the desired IP addresses.
+
```
Comment (frankbu) [inline content/en/faq/applications/elasticsearch.md]: Is all the following yaml needed? It's not clear to me what I'm supposed to do with it. If you include them, you should lead in with a sentence something like: "The following headless Service and StatefulSet is an example Elasticsearch configuration that will work in Istio.

Also why are there two separate snippets, below, showing the StatefulSet configuration?

File: content/en/faq/applications/elasticsearch.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  ports:
+    - port: 9200
+      name: http-rest
+    - port: 9300
+      name: inter-node
+{{< /text >}}
+
+1. `Statefulset` Definition
+
+Make sure ports are specified in the `Statefulset` definition as shown in the snippet below:
```
Comment (frankbu) [inline content/en/faq/applications/elasticsearch.md]: Is still true after https://github.com/istio/istio.io/pull/4859?

**Comment by mbanikazemi** [comment]
@linsun so this was (and is) a bug in support of headless services (stateful sets) with http ports. It turns out there was work being done on it while I was working one it. What got merged today still has one piece incorrectly that I have submitted as a PR: 16888

File: content/en/faq/applications/elasticsearch.md (hunk 4 above)
Comment (geeknoid) [reply content/en/faq/applications/elasticsearch.md]: Need a similar change on other blocks of yaml.

File: content/en/faq/applications/elasticsearch.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+needed.
+
+In order to run Elasticsearch with Istio one can explicitly set the
+`network.bind_host` to `0.0.0.0` or `localhost` (`127.0.0.1`) and
+`network.publish_host` to the pod IP to avoid any configuration issue.
+Refer to: [Network Settings for
+Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html#modules-network)
+for more information.
+
+With some configurations, ES can be deployed and work correctly
```
[the remark this answers is not in the document]
Comment (mbanikazemi) [reply content/en/faq/applications/elasticsearch.md]: Done.

File: content/en/faq/applications/elasticsearch.md (hunk 4 above)
Comment (mbanikazemi) [reply content/en/faq/applications/elasticsearch.md]: Done.

File: content/en/faq/applications/elasticsearch.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbanikazemi) [reply content/en/faq/applications/elasticsearch.md]: Edited.

File: content/en/faq/applications/elasticsearch.md (hunk 1 above)
Comment (mbanikazemi) [reply content/en/faq/applications/elasticsearch.md]: Done.

File: content/en/faq/applications/elasticsearch.md (hunk 2 above)
Comment (mbanikazemi) [reply content/en/faq/applications/elasticsearch.md]: Edited.

File: content/en/faq/applications/elasticsearch.md (hunk 6 above)
Comment (mbanikazemi) [reply content/en/faq/applications/elasticsearch.md]: Not any more. removed.

File: content/en/faq/applications/elasticsearch.md (hunk 5 above)
Comment (mbanikazemi) [reply content/en/faq/applications/elasticsearch.md]: OK.

File: content/en/faq/applications/index.md [hunk 8]
```diff
@@ -0,0 +1,10 @@
+---
+title: Applications FAQ
+linktitle: Applications
+description: Application Specific Q & A.
+weight: 30
+layout: faq
+icon: faq
+aliases:
+  - /help/faq/application
```
[the remark this answers is not in the document]
Comment (mbanikazemi) [reply content/en/faq/applications/index.md]: Done. Thanks for the correction.

File: .spelling [hunk 9]
```diff
@@ -480,6 +480,7 @@ SPIFFE
 SPIFFE-compliant
 SREs
 Stackdriver
+StatefulSet
```
[the remark this answers is not in the document]
Comment (mbanikazemi) [reply .spelling]: Done.

**Comment by geeknoid** [comment]
@rshriram Can you look at this please?

**Comment by frankbu** [comment]
@mbanikazemi Maybe it's just me, because I've never really tried to deploy elasticsearch, but this entry seems to be saying much more than necessary, which is just confusing IMO. There's also nothing in the text about why elasticsearch needs to be a headless service, but then you show it in your example. Could this entry be cut down to something shorter? What important information is missing if the entry were just the following?
```
---
title: Can I run Elasticsearch inside an Istio mesh?
description: Elasticsearch configuration required for use with Istio
weight: 50
keywords: [elasticsearch]
---

There are two Elasticsearch configuration parameters that need to be
set appropriately to run Elasticsearch with Istio:
`network.bind_host` and `network.publish_host`. By default, these
parameters are set to the `network.host` parameter. If `network.host`
is set to `0.0.0.0`, Elasticsearch will most likely pick up the pod IP
as the publishing address and no further configuration will be
needed.

If the default configuration does not work, you can set the
`network.bind_host` to `0.0.0.0` or `localhost` (`127.0.0.1`) and
`network.publish_host` to the pod IP. For example:

{{< text yaml >}}
      ...
      containers:
      - name: elasticsearch
        image: docker.elastic.co/elasticsearch/elasticsearch:7.2.0
        env:
          - name: network.bind_host
            value: 127.0.0.1
          - name: network.publish_host
            valueFrom:
              fieldRef:
                fieldPath: status.podIP
          ...
{{< /text >}}

Refer to [Network Settings for Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html#modules-network)
for more information.
```

**Comment by mbanikazemi** [comment]
@frankbu can you be specific and point to the info that you think is not needed?

**Comment by frankbu** [comment]
@mbanikazemi Everything I left out of my proposed version, above, I think may not be needed. The basic problem is making sure the pod IPs are reachable. Explaining how to configure that and giving a link to the elasticserach docs seems to be enough, unless I'm not understanding it right. If something else of what I left out is needed, please explain why and what is the minimum extra that needs to be said?

**Pull request merged**